File: scripts/run_simulation_batch.py
#!/usr/bin/env python3
"""
Batch Simulation Runner
Execute ERH simulations with configurable parameters and save results to JSON/CSV.
Supports parallel execution via multiprocessing for CI/CD pipelines.
Modes: judge (default) = judge-based ERH analysis; abm = Agent-Based Model simulation.
"""

import argparse
import json
import multiprocessing
import os
import sys
import time
from datetime import datetime
from pathlib import Path

# Ensure simulation module is in path
project_root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
sys.path.insert(0, project_root)

# Also add erh_core if it exists
erh_core_path = os.path.join(project_root, "erh_core")
if os.path.exists(erh_core_path) and erh_core_path not in sys.path:
    sys.path.insert(0, erh_core_path)

from simulation.core import (
    generate_world,
    BiasedJudge,
    evaluate_judgement,
    select_ethical_primes,
    compute_Pi_and_error,
    analyze_error_growth,
)
try:
    from erh_core.core.output_writer import save_json_result
except ImportError:
    from simulation.core.output_writer import save_json_result  # type: ignore


def _run_single_trial_abm(trial_id: int, config: dict) -> dict:
    """Worker for ABMSimulator mode: run one ABM trial and return metrics."""
    try:
        from erh_core.core.abm_simulator import ABMSimulator
    except ImportError:
        from simulation.core import ABMSimulator  # type: ignore
    n_agents = config.get("n_agents", 50)
    n_steps = config.get("steps", 100)
    sim = ABMSimulator(num_agents=n_agents)
    results = sim.run_simulation(num_time_steps=n_steps)
    erh_history = results.get("erh_history", [])
    final_error = 0.0
    quantum_energy = 0.0
    evs = 0.0
    if erh_history:
        last = erh_history[-1]
        analysis = last.get("analysis", {})
        E_x = last.get("E_x", [])
        if E_x is not None and len(E_x) > 0:
            import numpy as np
            final_error = float(np.mean(np.abs(E_x)))
        erh_satisfied = analysis.get("erh_satisfied", False)
        alpha = analysis.get("estimated_exponent", 0.5)
        stability = 1.0 if erh_satisfied else max(0, 0.5 - alpha)
        evs = stability * 0.9  # Placeholder; full EVS needs fairness, polarization
    return {
        "id": trial_id,
        "final_error": final_error,
        "social_tension": quantum_energy,
        "evs": evs,
        "erh_history_len": len(erh_history),
    }


def _run_single_config(config: dict, output_dir: str) -> str:
    """Worker function for a single simulation run (used by multiprocessing)."""
    sim_id = config.get("id", 0)
    num_actions = config.get("num_actions", 1000)
    complexity_dist = config.get("complexity_dist", "zipf")
    seed = config.get("seed", 42 + sim_id)

    start_time = time.time()
    actions = generate_world(
        num_actions=num_actions,
        complexity_dist=complexity_dist,
        random_seed=seed,
    )
    judge = BiasedJudge(bias_strength=0.2, noise_scale=0.1)
    evaluate_judgement(actions, judge, tau=0.3)
    primes = select_ethical_primes(actions, importance_quantile=0.9)
    Pi_x, B_x, E_x, x_vals = compute_Pi_and_error(primes, X_max=100)
    analysis = analyze_error_growth(E_x, x_vals)
    duration = time.time() - start_time
    mistakes = sum(1 for a in actions if a.mistake_flag)

    result = {
        "timestamp": datetime.now().isoformat(),
        "config": {
            "id": sim_id,
            "num_actions": num_actions,
            "complexity_dist": complexity_dist,
            "seed": seed,
        },
        "metrics": {
            "duration_seconds": duration,
            "total_actions": len(actions),
            "mistakes_count": mistakes,
            "mistake_rate": mistakes / len(actions),
            "ethical_primes_count": len(primes),
            "erh_satisfied": bool(analysis.get("erh_satisfied", False)),
            "estimated_exponent": analysis.get("estimated_exponent"),
        },
    }
    path = save_json_result(result, output_dir)
    stability = 1.0 - (mistakes / len(actions)) if actions else 0.0
    return f"Sim {sim_id} completed. Stability: {stability:.3f} -> {path}"


def run_simulation(args):
    """Run a single simulation configuration (legacy / non-parallel path)."""
    print(
        f"Starting simulation: N={args.num_actions}, Dist={args.complexity_dist}, Seed={args.seed}"
    )
    start_time = time.time()

    actions = generate_world(
        num_actions=args.num_actions,
        complexity_dist=args.complexity_dist,
        random_seed=args.seed,
    )
    judge = BiasedJudge(bias_strength=0.2, noise_scale=0.1)
    evaluate_judgement(actions, judge, tau=0.3)
    primes = select_ethical_primes(actions, importance_quantile=0.9)
    Pi_x, B_x, E_x, x_vals = compute_Pi_and_error(primes, X_max=100)
    analysis = analyze_error_growth(E_x, x_vals)
    duration = time.time() - start_time
    mistakes = sum(1 for a in actions if a.mistake_flag)

    result = {
        "timestamp": datetime.now().isoformat(),
        "config": {
            "num_actions": args.num_actions,
            "complexity_dist": args.complexity_dist,
            "seed": args.seed,
        },
        "metrics": {
            "duration_seconds": duration,
            "total_actions": len(actions),
            "mistakes_count": mistakes,
            "mistake_rate": mistakes / len(actions),
            "ethical_primes_count": len(primes),
            "erh_satisfied": bool(analysis.get("erh_satisfied", False)),
            "estimated_exponent": analysis.get("estimated_exponent"),
        },
    }
    return result


def main():
    parser = argparse.ArgumentParser(description="ERH Batch Simulation Runner")
    parser.add_argument(
        "--mode",
        type=str,
        default="judge",
        choices=["judge", "abm"],
        help="Simulation mode: judge (default) or abm (Agent-Based Model)",
    )
    parser.add_argument(
        "--num-actions", type=int, default=1000, help="Number of actions (judge mode)"
    )
    parser.add_argument(
        "--agents", type=int, default=50, help="Number of agents (abm mode)"
    )
    parser.add_argument(
        "--steps", type=int, default=100, help="Time steps (abm mode)"
    )
    parser.add_argument(
        "--trials", type=int, default=4, help="Number of trials (abm mode)"
    )
    parser.add_argument(
        "--complexity-dist",
        type=str,
        default="zipf",
        choices=["zipf", "uniform", "power_law"],
    )
    parser.add_argument("--seed", type=int, default=42, help="Random seed")
    parser.add_argument(
        "--output-dir",
        type=str,
        default="results",
        help="Output directory",
    )
    parser.add_argument(
        "--output",
        type=str,
        default=None,
        help="Path for aggregated JSON output (abm mode or when --instances>1). Default: simulation/output/batch_results.json",
    )
    parser.add_argument(
        "--instances",
        type=int,
        default=1,
        help="Number of parallel simulation instances (multiprocessing)",
    )
    parser.add_argument(
        "--configs",
        type=str,
        default=None,
        help="Path to JSON config file (list of configs). If not set, uses --num-actions/--complexity-dist/--seed with --instances runs.",
    )
    parser.add_argument(
        "--real-data-only",
        action="store_true",
        help="Run only empirical validation (COMPAS, Adult) and skip judge/ABM simulation.",
    )

    args = parser.parse_args()

    # Real-data-only mode: run empirical validation script
    if args.real_data_only:
        run_emp = project_root / "scripts" / "run_empirical_validation.py"
        if run_emp.exists():
            import subprocess
            out_dir = args.output_dir or os.path.join(project_root, "simulation", "output")
            result = subprocess.run(
                [sys.executable, str(run_emp), "--output", str(Path(out_dir) / "real_world_results.json")],
                cwd=str(project_root),
                env={**os.environ, "PYTHONPATH": f"{project_root}:{os.path.join(project_root, 'erh_core')}"},
            )
            sys.exit(result.returncode)
        else:
            print("run_empirical_validation.py not found")
            sys.exit(1)

    # ABM mode
    if args.mode == "abm":
        config = {"n_agents": args.agents, "steps": args.steps}
        start_time = time.time()
        with multiprocessing.Pool() as pool:
            results = pool.starmap(_run_single_trial_abm, [(i, config) for i in range(args.trials)])
        duration = time.time() - start_time
        output_path = args.output or os.path.join(project_root, "simulation", "output", "batch_results.json")
        Path(output_path).parent.mkdir(parents=True, exist_ok=True)
        with open(output_path, "w") as f:
            json.dump({"config": config, "results": results, "duration_seconds": duration}, f, indent=2)
        print(f"ABM batch completed in {duration:.2f}s ({args.trials} trials) -> {output_path}")
        return

    if args.instances <= 1 and args.configs is None:
        # Single-run path (backward compatible)
        try:
            result = run_simulation(args)
            path = save_json_result(result, args.output_dir)
            print(f"Results saved to {path}")
        except Exception as e:
            print(f"Error running simulation: {e}")
            sys.exit(1)
        return

    # Parallel batch path
    os.makedirs(args.output_dir, exist_ok=True)
    if args.configs and os.path.isfile(args.configs):
        with open(args.configs) as f:
            configs = json.load(f)
        if not isinstance(configs, list):
            configs = [configs]
        for i, c in enumerate(configs):
            if "id" not in c:
                c["id"] = i
    else:
        configs = [
            {
                "id": i,
                "num_actions": args.num_actions,
                "complexity_dist": args.complexity_dist,
                "seed": args.seed + i,
            }
            for i in range(args.instances)
        ]

    n_workers = min(args.instances, len(configs), multiprocessing.cpu_count())
    start_time = time.time()
    with multiprocessing.Pool(processes=n_workers) as pool:
        results = pool.starmap(
            _run_single_config,
            [(cfg, args.output_dir) for cfg in configs],
        )
    elapsed = time.time() - start_time
    print(f"Batch completed in {elapsed:.2f}s ({len(configs)} simulations)")
    for r in results:
        print(r)


if __name__ == "__main__":
    main()

File: julia/scripts/run_simulation_batch.jl
#!/usr/bin/env julia
"""
Batch Simulation Runner (Julia)
Execute ERH simulations with configurable parameters and save results to JSON/CSV.
Modes: judge (default) = judge-based ERH analysis; abm = Agent-Based Model simulation.

Usage:
    julia --project=julia julia/scripts/run_simulation_batch.jl [options]
    julia --project=julia julia/scripts/run_simulation_batch.jl --smoke
"""

push!(LOAD_PATH, joinpath(@__DIR__, "..", "src"))
using ERH
using JSON3
using Random
using Dates
using Statistics

# ---------------------------------------------------------------------------
# Argument parsing
# ---------------------------------------------------------------------------
function parse_args(argv=ARGS)
    args = Dict{String,Any}(
        "mode"            => "judge",
        "num_actions"     => 1000,
        "agents"          => 50,
        "steps"           => 100,
        "trials"          => 4,
        "complexity_dist" => "zipf",
        "seed"            => 42,
        "output_dir"      => "results",
        "output"          => nothing,
        "instances"       => 1,
        "configs"         => nothing,
        "smoke"           => false,
    )
    i = 1
    while i <= length(argv)
        a = argv[i]
        if a == "--smoke"
            args["smoke"] = true
        elseif a == "--mode" && i < length(argv)
            i += 1; args["mode"] = argv[i]
        elseif a == "--num-actions" && i < length(argv)
            i += 1; args["num_actions"] = parse(Int, argv[i])
        elseif a == "--agents" && i < length(argv)
            i += 1; args["agents"] = parse(Int, argv[i])
        elseif a == "--steps" && i < length(argv)
            i += 1; args["steps"] = parse(Int, argv[i])
        elseif a == "--trials" && i < length(argv)
            i += 1; args["trials"] = parse(Int, argv[i])
        elseif a == "--complexity-dist" && i < length(argv)
            i += 1; args["complexity_dist"] = argv[i]
        elseif a == "--seed" && i < length(argv)
            i += 1; args["seed"] = parse(Int, argv[i])
        elseif a == "--output-dir" && i < length(argv)
            i += 1; args["output_dir"] = argv[i]
        elseif a == "--output" && i < length(argv)
            i += 1; args["output"] = argv[i]
        elseif a == "--instances" && i < length(argv)
            i += 1; args["instances"] = parse(Int, argv[i])
        elseif a == "--configs" && i < length(argv)
            i += 1; args["configs"] = argv[i]
        end
        i += 1
    end
    return args
end

# ---------------------------------------------------------------------------
# ERH simulation helpers (pure-Julia reimplementation)
# ---------------------------------------------------------------------------

"""Generate a synthetic action world with the given complexity distribution."""
function generate_world(; num_actions::Int, complexity_dist::String="zipf", seed::Int=42)
    rng = MersenneTwister(seed)
    actions = []
    for i in 1:num_actions
        c = if complexity_dist == "zipf"
            clamp(round(Int, 100 / (rand(rng) * 99 + 1)), 1, 100)
        elseif complexity_dist == "uniform"
            rand(rng, 1:100)
        else  # power_law
            clamp(round(Int, 100 * rand(rng)^2), 1, 100)
        end
        w = rand(rng)          # moral weight in [0,1]
        v = rand(rng)          # ground-truth value
        push!(actions, (id=i, c=c, w=w, v=v, mistake_flag=false))
    end
    return actions
end

"""Apply biased judge evaluation, setting mistake_flag on each action."""
function evaluate_judgement!(actions; bias_strength=0.2, noise_scale=0.1, tau=0.3, seed=0)
    rng = MersenneTwister(seed + 1000)
    result = []
    for a in actions
        noise = randn(rng) * noise_scale
        judgment = clamp(a.v + bias_strength * (0.5 - a.v) + noise, 0.0, 1.0)
        mistake = abs(judgment - a.v) > tau
        push!(result, merge(a, (mistake_flag=mistake,)))
    end
    return result
end

"""Select ethical primes: top importance_quantile by moral weight."""
function select_ethical_primes(actions; importance_quantile=0.9)
    threshold = quantile([a.w for a in actions], importance_quantile)
    return filter(a -> a.w >= threshold, actions)
end

"""Compute Pi(x), B(x), E(x) and x_vals over [1, X_max]."""
function compute_Pi_and_error(primes; X_max=100)
    x_vals = collect(1:X_max)
    Pi_x = [count(p -> p.c <= x, primes) for x in x_vals]
    # Li(x) approximation as baseline B(x)
    B_x = [x <= 1 ? 1.0 : x / log(x) for x in x_vals]
    E_x = [Pi_x[i] - B_x[i] for i in eachindex(x_vals)]
    return Pi_x, B_x, E_x, x_vals
end

"""Estimate alpha (growth exponent) from |E(x)| in log-log space."""
function analyze_error_growth(E_x, x_vals)
    valid = [(log(x), log(abs(e) + 1e-9)) for (x, e) in zip(x_vals, E_x) if x > 1 && abs(e) > 1e-9]
    if length(valid) < 5
        return Dict("erh_satisfied" => false, "estimated_exponent" => 0.5)
    end
    xs = [v[1] for v in valid]
    ys = [v[2] for v in valid]
    n = length(xs)
    x_mean, y_mean = mean(xs), mean(ys)
    alpha = sum((xs .- x_mean) .* (ys .- y_mean)) / (sum((xs .- x_mean).^2) + 1e-12)
    erh_satisfied = 0.4 <= alpha <= 0.65
    return Dict("erh_satisfied" => erh_satisfied, "estimated_exponent" => alpha)
end

# ---------------------------------------------------------------------------
# ABM simulation helper
# ---------------------------------------------------------------------------
function run_abm_trial(trial_id::Int, n_agents::Int, n_steps::Int, seed::Int)
    rng = MersenneTwister(seed + trial_id * 37)
    # Simple ABM: agents interact, ethical drift accumulates
    state = rand(rng, n_agents)
    erh_history = []
    for t in 1:n_steps
        # Pairwise interactions
        for _ in 1:n_agents
            i, j = rand(rng, 1:n_agents, 2)
            state[i] = clamp(state[i] + 0.01 * (state[j] - state[i]) + 0.01 * randn(rng), 0, 1)
        end
        err = mean(abs.(state .- 0.5))
        push!(erh_history, Dict("t" => t, "mean_error" => err))
    end
    final_error = isempty(erh_history) ? 0.0 : erh_history[end]["mean_error"]
    evs = max(0.0, 1.0 - final_error)
    return Dict(
        "id"               => trial_id,
        "final_error"      => final_error,
        "social_tension"   => 0.0,
        "evs"              => evs,
        "erh_history_len"  => length(erh_history),
    )
end

# ---------------------------------------------------------------------------
# Single simulation run
# ---------------------------------------------------------------------------
function run_single_simulation(sim_id::Int, num_actions::Int, complexity_dist::String, seed::Int)
    t0 = time()
    actions = generate_world(num_actions=num_actions, complexity_dist=complexity_dist, seed=seed)
    actions = evaluate_judgement!(actions; bias_strength=0.2, noise_scale=0.1, tau=0.3, seed=seed)
    primes  = select_ethical_primes(actions; importance_quantile=0.9)
    Pi_x, B_x, E_x, x_vals = compute_Pi_and_error(primes; X_max=100)
    analysis = analyze_error_growth(E_x, x_vals)
    duration = time() - t0
    mistakes = count(a -> a.mistake_flag, actions)
    return Dict(
        "timestamp" => string(now()),
        "config"    => Dict(
            "id"              => sim_id,
            "num_actions"     => num_actions,
            "complexity_dist" => complexity_dist,
            "seed"            => seed,
        ),
        "metrics"   => Dict(
            "duration_seconds"      => duration,
            "total_actions"         => length(actions),
            "mistakes_count"        => mistakes,
            "mistake_rate"          => mistakes / length(actions),
            "ethical_primes_count"  => length(primes),
            "erh_satisfied"         => get(analysis, "erh_satisfied", false),
            "estimated_exponent"    => get(analysis, "estimated_exponent", nothing),
        ),
    )
end

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
function main()
    args = parse_args()

    if args["smoke"]
        println("[smoke] run_simulation_batch: minimal run (2 actions, 1 instance)")
        args["num_actions"] = 2
        args["instances"]   = 1
        args["trials"]      = 1
        args["steps"]       = 2
        args["agents"]      = 2
    end

    mode       = args["mode"]
    output_dir = args["output_dir"]
    mkpath(output_dir)

    if mode == "abm"
        n_agents = args["agents"]
        n_steps  = args["steps"]
        n_trials = args["trials"]
        seed     = args["seed"]
        println("[batch] ABM mode: agents=$n_agents, steps=$n_steps, trials=$n_trials")
        t0 = time()
        results = [run_abm_trial(i, n_agents, n_steps, seed) for i in 0:(n_trials - 1)]
        duration = time() - t0
        output_path = something(args["output"], joinpath(output_dir, "batch_results.json"))
        mkpath(dirname(output_path))
        open(output_path, "w") do io
            JSON3.pretty(io, Dict(
                "config"           => Dict("n_agents" => n_agents, "steps" => n_steps),
                "results"          => results,
                "duration_seconds" => duration,
            ))
        end
        println("[batch] ABM batch completed in $(round(duration; digits=2))s ($n_trials trials) -> $output_path")
        return
    end

    # Judge mode
    configs = if args["configs"] !== nothing && isfile(args["configs"])
        raw = JSON3.read(read(args["configs"], String))
        isa(raw, AbstractArray) ? collect(raw) : [raw]
    else
        n = args["instances"]
        [Dict(
            "id"              => i,
            "num_actions"     => args["num_actions"],
            "complexity_dist" => args["complexity_dist"],
            "seed"            => args["seed"] + i,
        ) for i in 0:(n - 1)]
    end

    println("[batch] Judge mode: $(length(configs)) simulation(s)")
    t0 = time()
    for cfg in configs
        sim_id    = get(cfg, "id", 0)
        n_actions = get(cfg, "num_actions", args["num_actions"])
        cdist     = get(cfg, "complexity_dist", args["complexity_dist"])
        s         = get(cfg, "seed", args["seed"])
        println("  [sim $sim_id] N=$n_actions dist=$cdist seed=$s ...")
        result = run_single_simulation(sim_id, n_actions, cdist, s)
        out_path = joinpath(output_dir, "sim_$(sim_id)_$(Dates.format(now(), "yyyymmdd_HHMMSS")).json")
        open(out_path, "w") do io
            JSON3.pretty(io, result)
        end
        mistakes  = result["metrics"]["mistakes_count"]
        total     = result["metrics"]["total_actions"]
        stability = 1.0 - mistakes / max(total, 1)
        println("  [sim $sim_id] Stability: $(round(stability; digits=3)) -> $out_path")
    end
    elapsed = time() - t0
    println("[batch] Batch completed in $(round(elapsed; digits=2))s ($(length(configs)) simulations)")
end

main()
